src/main/java/gregtech/api/util/GTTransferUtils.java

```diff
@@ -0,0 +1,250 @@
+                if (amount > 0) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // perform the merge.
+        items.forEach(stack -> insertItem(handler, stack, false));
+        return true;
+    }
+
+    /**
+     * Simulates the insertion of fluid into a target fluid handler, then optionally performs the insertion.
+     * <br /><br />
+     * Simulating will not modify any of the input parameters. Insertion will either succeed completely, or fail
+     * without modifying anything.
+     * This method should be called with {@code simulate} {@code true} first, then {@code simulate} {@code false},
+     * only if it returned {@code true}.
+     *
+     * @param fluidHandler the target inventory
+     * @param simulate     whether to simulate ({@code true}) or actually perform the insertion ({@code false})
+     * @param fluidStacks  the items to insert into {@code fluidHandler}.
+     * @return {@code true} if the insertion succeeded, {@code false} otherwise.
+     */
+    public static boolean addFluidsToFluidHandler(IMultipleTankHandler fluidHandler,
+                                                  boolean simulate,
+                                                  List<FluidStack> fluidStacks) {
+        if (simulate) {
+            OverlayedFluidHandler overlayedFluidHandler = new OverlayedFluidHandler(fluidHandler);
+            HashMap<FluidKey, Integer> fluidKeyMap = GTHashMaps.fromFluidCollection(fluidStacks);
+            for (Map.Entry<FluidKey, Integer> entry : fluidKeyMap.entrySet()) {
+                int amountToInsert = entry.getValue();
+                int inserted = overlayedFluidHandler.insertStackedFluidKey(entry.getKey(), amountToInsert);
+                if (inserted != amountToInsert) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        fluidStacks.forEach(fluidStack -> fluidHandler.fill(fluidStack, true));
+        return true;
+    }
+
+    /**
+     * Inserts items by trying to fill slots with the same item first, and then fill empty slots.
+     */
+    public static ItemStack insertItem(IItemHandler handler, ItemStack stack, boolean simulate) {
+        if (handler == null || stack.isEmpty()) {
+            return stack;
+        }
+        if (!stack.isStackable()) {
+            return insertToEmpty(handler, stack, simulate);
+        }
+
+        List<Integer> emptySlots = new ArrayList<>();
+        int slots = handler.getSlots();
+
+        for (int i = 0; i < slots; i++) {
+            ItemStack slotStack = handler.getStackInSlot(i);
+            if (slotStack.isEmpty()) {
+                emptySlots.add(i);
+            }
+            if (areStackable(stack, slotStack)) {
+                stack = handler.insertItem(i, stack, simulate);
+                if (stack.isEmpty()) {
+                    return ItemStack.EMPTY;
+                }
+            }
+        }
```

Comment:
Forge's `ItemHandlerHelper` seems to involve some of these methods, would be great to check if the logic is the same and/or even use those methods instead.

Reply:
I removed one methods since it matches the logic of a existing method. The other two are very similar, but are a tiny bit more effecient.